zk-gadget/src/mock.rs

```diff
@@ -0,0 +1,399 @@
+        // forget rx so that it doesn't get dropped
+        core::mem::forget(rx);
+        let header = self
+            .externalities
+            .lock()
+            .execute_with(|| System::finalize());
+        let summary = FinalizeSummary::<Block> {
+            finalized: vec![header.hash()],
+            header,
+            stale_heads: vec![],
+        };
+        let notification = FinalityNotification::from_summary(summary, tx.clone());
+        Some(notification)
+    }
+
+    async fn get_next_block_import_notification(&self) -> Option<BlockImportNotification<Block>> {
+        futures::future::pending().await
+    }
+}
+
+impl ProvideRuntimeApi<Block> for MockClient {
+    type Api = Runtime;
+    fn runtime_api(&self) -> ApiRef<Self::Api> {
+        ApiRef::from(self.runtime)
+    }
+}
+
+sp_api::mock_impl_runtime_apis! {
+  impl pallet_jobs_rpc_runtime_api::JobsApi<Block, AccountId32> for Runtime {
+    fn query_jobs_by_validator(
+        validator: AccountId32,
+    ) -> Result<Vec<RpcResponseJobsData<AccountId32>>, String> {
+        Jobs::query_jobs_by_validator(validator)
+    }
+
+    fn query_job_by_id(job_key: JobKey, job_id: JobId) -> Option<RpcResponseJobsData<AccountId32>> {
+        Jobs::submitted_jobs(job_key, job_id).map(|job| {
+            if !job.job_type.is_phase_one() {
+                let result = Jobs::known_results(
+                    job.job_type.get_previous_phase_job_key().unwrap(),
+                    job.job_type.clone().get_phase_one_id().unwrap(),
+                )
+                .unwrap();
+
+                RpcResponseJobsData {
+                    job_id,
+                    job_type: job.job_type,
+                    participants: result.participants(),
+                    threshold: result.threshold(),
+                    key: Some(result.result),
+                }
+            } else {
+                RpcResponseJobsData {
+                    job_id,
+                    job_type: job.job_type,
+                    participants: None,
+                    threshold: None,
+                    key: None,
+                }
+            }
+        })
+    }
+
+    fn query_phase_one_by_id(
+        job_key: JobKey,
+        job_id: JobId,
+    ) -> Option<RpcResponsePhaseOneResult<AccountId32>> {
+        Jobs::known_results(job_key, job_id).map(|result| RpcResponsePhaseOneResult {
+            owner: result.owner,
+            result: result.result,
+            permitted_caller: result.permitted_caller,
+            key_type: result.key_type,
+            job_type: result.job_type,
+        })
+    }
+
+    fn query_next_job_id() -> JobId {
+        Jobs::next_job_id()
+    }
+  }
+}
+
+pub fn to_account_id32(id: u8) -> AccountId32 {
+    AccountId32::new([id; 32])
+}
+
+pub fn alloc_available_ip_addr() -> SocketAddr {
+    // use port 0 to let the OS allocate a port
+    SocketAddr::new([127, 0, 0, 1].into(), 0)
```

Comment:
Are we sure using `0` will automatically allocate a new port? In my experience, the allocation doesn't happen until we call `bind` on a port. So, if a client uses this socket address, it won't be able to connect since 127.0.0.1:0 is not an actual port.